test/e2e/proxy_test.go

```diff
@@ -160,6 +181,160 @@ func TestProxyFlow(t *testing.T) {
 				}
 			})
 
+			t.Run("use proxy get a HAS Application CR in the user appstudio namespace via proxy API and use websocket to verify", func(t *testing.T) {
+				// Start a new websocket watcher which watches for Application CRs in the user's namespace
+				w := newWsWatcher(t, *user, user.compliantUsername, hostAwait.APIProxyURL)
+				closeConnection := w.Start()
+				defer closeConnection()
+				proxyCl, err := hostAwait.CreateAPIProxyClient(t, user.token, hostAwait.APIProxyURL)
+				require.NoError(t, err)
+
+				// Retrieve the application resources multiple times for the same user to make sure the proxy cache kicks in.
+				for i := 0; i < 2; i++ {
+					// given
+					applicationName := user.getApplicationName(i)
+					// Get Application
+					proxyApp := user.getApplication(t, proxyCl, applicationName)
+
+					// Double check that the Application using a regular client (non-proxy)
+					wsApp := &appstudiov1.Application{}
+					namespacedName := types.NamespacedName{Namespace: tenantNsName(user.compliantUsername), Name: applicationName}
+					err = user.expectedMemberCluster.Client.Get(context.TODO(), namespacedName, wsApp)
+					require.NoError(t, err)
+					require.NotEmpty(t, wsApp)
+					assert.Equal(t, proxyApp, wsApp)
+				}
+			})
+
+			t.Run("use proxy to update a HAS Application CR in the user appstudio namespace via proxy API and use websocket to watch it updated", func(t *testing.T) {
+				// Start a new websocket watcher which watches for Application CRs in the user's namespace
+				w := newWsWatcher(t, *user, user.compliantUsername, hostAwait.APIProxyURL)
+				closeConnection := w.Start()
+				defer closeConnection()
+				proxyCl, err := hostAwait.CreateAPIProxyClient(t, user.token, hostAwait.APIProxyURL)
+				require.NoError(t, err)
+
+				// Retrieve the application resources multiple times for the same user to make sure the proxy cache kicks in.
+				for i := 0; i < 2; i++ {
```

Comment:
Instead of re-creating all this infrastructure for every sub test (proxy client, websocket watcher, etc) I would just encapsulate the sub-tests to reproduce the whole flow:

```
t.Run("use proxy to create a HAS Application CR in the user appstudio namespace via proxy API and use websocket to watch it created", func(t *testing.T) {
    // Create a proxy client, start a watcher, etc
    // Create an app, watch it created, check you can get it via proxy
    ...
    t.Run("use proxy to update a HAS Application...", func(t *testing.T) {
        // Reuse the proxy client, watcher, etc created above
        // Update the app
        ...
       t.Run("use proxy to delete a HAS Application...", func(t *testing.T) {
            // Reuse the proxy client, watcher, etc created above
            // Delete the app
           ...
        }
    }
}
```

Or alternatively:
```
// Create a proxy client, start a watcher, etc
...
t.Run("use proxy to create a HAS Application CR in the user appstudio namespace via proxy API and use websocket to watch it created", func(t *testing.T) {
    // Reuse the proxy client, watcher, etc created above
    // Create an app, watch it created, check you can get it via proxy
    ...
}

t.Run("use proxy to update the HAS Application created above...", func(t *testing.T) {
    // Reuse the proxy client, watcher, etc created above
    // Update the app
    ...
}

t.Run("use proxy to delete the HAS Application created above...", func(t *testing.T) {
    // Reuse the proxy client, watcher, etc created above
    // Delete the app
    ...
}
```

Also, I think we don't have to retry every operation multiple times to test the cache. It's enough to do it just when we are creating the app in the very first test.

Reply:
I have rewritten code to use the second approach. Resolving.